cmd/agent/main.go

```diff
@@ -24,54 +26,57 @@ func main() {
 		os.Exit(1)
 	}
 
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
 	metricsStore := storage.NewMetricsCollection()
 	metricsRepo := repositories.NewAgentRepository(metricsStore, agentCfg.Log)
-	metricsService := agent.NewAgentService(metricsRepo, agentCfg.Log)
+	metricsService := agent.NewAgentService(metricsRepo, agentCfg.Log, &agentCfg.Key)
+	// Create worker pool with the specified rate limit
+	workerPool := worker.NewWorkerPool(agentCfg.RateLimit)
 
 	pollTicker := time.NewTicker(agentCfg.PollInterval)
 	defer pollTicker.Stop()
 	reportTicker := time.NewTicker(agentCfg.ReportInterval)
 	defer reportTicker.Stop()
 
-	originalReportInterval := agentCfg.ReportInterval
-
-	for {
-		select {
-		case <-pollTicker.C:
-			if err = metricsService.MetricsService.Collect(); err != nil {
-				agentCfg.Log.ErrorContext(context.Background(),
-					"failed to collect the metrics: ",
-					helpers.ErrAttr(err))
-			}
-		case <-reportTicker.C:
-			retries := 3
-			backoffIntervals := []time.Duration{1 * time.Second, 3 * time.Second, 5 * time.Second}
-			success := false
-
-			for attempt := range backoffIntervals {
-				err = metricsService.MetricsService.Send(agentCfg.ServerAddr)
-				if err == nil {
-					metricsStore.Clear()
-					reportTicker.Reset(originalReportInterval)
-					success = true
-					break
-				}
+	// originalReportInterval := agentCfg.ReportInterval
 
-				agentCfg.Log.ErrorContext(context.Background(),
-					fmt.Sprintf("Attempt %d: failed to publish the metrics: %v", attempt+1, err),
-					helpers.ErrAttr(err))
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
+	go func() {
+		<-sigChan
+		cancel()
```

Comment:
If you want to implement graceful shutdown, you should wait for all components to stop before stopping the process: currently, after returning on the line 79 your process finishes independently from the worker pool or other goroutines.

You can use a wait group to achieve that